File: apps/orchestrator-api/src/qdrant_client.py
from typing import Optional
import httpx

from src.config import get_settings
from src.logging import get_logger

logger = get_logger("qdrant-client")


class QdrantClient:
    def __init__(self):
        settings = get_settings()
        self.host = settings.qdrant_host
        self.port = settings.qdrant_port
        self.api_key = settings.qdrant_api_key if settings.qdrant_api_key else None
        self.base_url = f"http://{self.host}:{self.port}"
        self._client = httpx.AsyncClient(timeout=30.0)
        self._headers = {"api-key": self.api_key} if self.api_key else {}
        logger.info("QdrantClient initialized", {"host": self.host, "port": self.port, "has_api_key": bool(self.api_key)})

    async def upsert(self, collection: str, point_id: str, vector: list[float], payload: dict) -> bool:
        await self._ensure_collection(collection, len(vector))
        url = f"{self.base_url}/collections/{collection}/points?wait=true"
        body = {
            "points": [
                {
                    "id": point_id,
                    "vector": vector,
                    "payload": payload,
                }
            ]
        }
        try:
            response = await self._client.put(url, json=body, headers=self._headers)
            response.raise_for_status()
            return True
        except Exception as e:
            logger.error("Qdrant upsert failed", {"error": str(e)})
            return False

    async def _ensure_collection(self, collection: str, vector_size: int):
        url = f"{self.base_url}/collections/{collection}"
        try:
            response = await self._client.get(url, headers=self._headers)
            if response.status_code == 404:
                create_url = f"{self.base_url}/collections/{collection}"
                create_body = {
                    "vectors": {
                        "size": vector_size,
                        "distance": "Cosine",
                    }
                }
                response = await self._client.put(create_url, json=create_body, headers=self._headers)
                response.raise_for_status()
                logger.info("Qdrant collection created", {"collection": collection, "vector_size": vector_size})
        except Exception as e:
            logger.warn("Qdrant collection check failed", {"error": str(e)})

    async def search(self, collection: str, vector: list[float], limit: int = 5, filter_payload: Optional[dict] = None) -> list[dict]:
        url = f"{self.base_url}/collections/{collection}/points/search"
        body = {
            "vector": vector,
            "limit": limit,
            "with_payload": True,
        }
        if filter_payload:
            body["filter"] = {
                "must": [{"key": k, "match": {"value": v}} for k, v in filter_payload.items()]
            }
        try:
            response = await self._client.post(url, json=body, headers=self._headers)
            response.raise_for_status()
            data = response.json()
            return [
                {
                    "id": p["id"],
                    "score": p["score"],
                    **p.get("payload", {}),
                }
                for p in data.get("result", [])
            ]
        except Exception as e:
            logger.error("Qdrant search failed", {"error": str(e)})
            return []

    async def delete(self, point_id: str, collection: str = "memories") -> bool:
        url = f"{self.base_url}/collections/{collection}/points/delete"
        body = {"points": [point_id]}
        try:
            response = await self._client.post(url, json=body, headers=self._headers)
            response.raise_for_status()
            return True
        except Exception as e:
            logger.error("Qdrant delete failed", {"error": str(e)})
            return False

    async def close(self):
        await self._client.aclose()


def create_qdrant_client() -> QdrantClient:
    return QdrantClient()

File: packages/memory/src/qdrant-client.ts
import { QdrantClient as QdrantSDK } from "@qdrant/js-client-rest";
import { VectorSearchResult, MemoryCategory } from "./types";
import { qdrantConfig } from "./config";
import { get_logger } from "./logging";

const logger = get_logger("qdrant-client");

const COLLECTION_NAME = "lira_memories";

export interface QdrantClient {
  upsert(id: string, vector: number[], payload: Record<string, unknown>): Promise<boolean>;
  search(vector: number[], limit?: number, categoryFilter?: MemoryCategory[]): Promise<VectorSearchResult[]>;
  delete(id: string): Promise<boolean>;
  deleteByCategory(category: MemoryCategory): Promise<boolean>;
  collectionExists(): Promise<boolean>;
  ensureCollection(vectorSize: number): Promise<void>;
  healthCheck(): Promise<boolean>;
}

export function createQdrantClient(): QdrantClient {
  const client = new QdrantSDK({
    url: qdrantConfig.url,
    apiKey: qdrantConfig.apiKey || undefined,
  });

  return {
    async upsert(id: string, vector: number[], payload: Record<string, unknown>): Promise<boolean> {
      try {
        await client.upsert(COLLECTION_NAME, {
          wait: true,
          points: [
            {
              id,
              vector,
              payload,
            },
          ],
        });
        return true;
      } catch (error) {
        logger.error("Failed to upsert vector", { id, error: String(error) });
        return false;
      }
    },

    async search(
      vector: number[],
      limit: number = 10,
      categoryFilter?: MemoryCategory[],
    ): Promise<VectorSearchResult[]> {
      try {
        const filter = categoryFilter
          ? {
              should: categoryFilter.map((cat) => ({
                key: "category",
                match: { value: cat },
              })),
            }
          : undefined;

        const results = await client.search(COLLECTION_NAME, {
          vector,
          limit,
          filter,
          with_payload: true,
        });

        return results.map((r) => ({
          id: String(r.id),
          score: r.score,
          payload: r.payload as Record<string, unknown>,
          category: (r.payload?.category as MemoryCategory) ?? undefined,
        }));
      } catch (error) {
        logger.error("Vector search failed", { error: String(error) });
        return [];
      }
    },

    async delete(id: string): Promise<boolean> {
      try {
        await client.delete(COLLECTION_NAME, {
          wait: true,
          points: [id],
        });
        return true;
      } catch (error) {
        logger.error("Failed to delete vector", { id, error: String(error) });
        return false;
      }
    },

    async deleteByCategory(category: MemoryCategory): Promise<boolean> {
      try {
        await client.delete(COLLECTION_NAME, {
          wait: true,
          filter: {
            must: [
              {
                key: "category",
                match: { value: category },
              },
            ],
          },
        });
        return true;
      } catch (error) {
        logger.error("Failed to delete vectors by category", { category, error: String(error) });
        return false;
      }
    },

    async collectionExists(): Promise<boolean> {
      try {
        const collections = await client.getCollections();
        return collections.collections.some((c) => c.name === COLLECTION_NAME);
      } catch {
        return false;
      }
    },

    async ensureCollection(vectorSize: number): Promise<void> {
      const exists = await this.collectionExists();
      if (exists) return;

      try {
        await client.createCollection(COLLECTION_NAME, {
          vectors: {
            size: vectorSize,
            distance: "Cosine",
          },
        });
        logger.info("Created Qdrant collection", { name: COLLECTION_NAME, vectorSize });
      } catch (error) {
        logger.error("Failed to create collection", { error: String(error) });
        throw error;
      }
    },

    async healthCheck(): Promise<boolean> {
      try {
        await client.getCollections();
        return true;
      } catch {
        return false;
      }
    },
  };
}
